File: Sources/ManageArmsCore/WriteGuard.swift
import Foundation

/// 触ってよい対象の限定。DESIGN.md 9 章。
///
/// 「触ってはいけないパスの列挙」ではなく「触ってよい対象の限定」にしてある。
/// 新しいリソース種別を足しても安全側に倒れる。
public enum WriteGuard {

    public enum Denial: Error, Equatable, CustomStringConvertible {
        /// 二重チェックに引っかかった。認証ファイル等。
        case deniedPath(String)
        /// symlink のリンク先が実体置き場の外を指している。自分が張ったものではない。
        case symlinkOutsideStore(String)
        /// 実体置き場でも退避ディレクトリでもない場所。
        case outsideManagedRoots(String)
        /// registry.json に載っていない。他ツールが入れたもの（外部管理）。
        case notInRegistry(String)
        /// 取得物が名乗った名前がパス要素として使えない（`..` / `/` を含む等）。
        case invalidName(String)
        /// 書き込み先までの既存ディレクトリに symlink が含まれる。
        case unsafeParent(String)

        public var description: String {
            switch self {
            case .deniedPath(let p):
                String(localized: "\(p) は保護対象です")
            case .symlinkOutsideStore(let p):
                String(localized: "\(p) は manage-arms が張った symlink ではありません")
            case .outsideManagedRoots(let p):
                String(localized: "\(p) は manage-arms の管理外です")
            case .notInRegistry(let n):
                String(localized: "\(n) は他のツールが管理しています。manage-arms からは変更できません")
            case .invalidName(let n):
                String(localized: "\(n) は名前として使えません（取得元の指定を確認してください）")
            case .unsafeParent(let p):
                String(localized: "\(p) の親ディレクトリに安全でないシンボリックリンクがあります")
            }
        }
    }

    /// ホワイトリストに通っても無条件で拒否する。
    /// ホワイトリストの実装ミス 1 つで到達しうる場所なので二重にする（9 章）。
    /// `~/.codex/auth.json` は認証トークンで、誤爆すると全ログインが飛ぶ。
    static let deniedNames: Set<String> = [
        "auth.json", "oauth_creds.json", "settings.json",
        "settings.local.json", ".claude.json", "config.toml", "mcp.json",
    ]
    static let deniedExtensions: Set<String> = ["sqlite", "sqlite-wal", "sqlite-shm"]

    /// 取得物が名乗った名前を、そのままパス要素に使ってよいか。**純粋関数**（10.1）。
    ///
    /// **名前は取得先リポジトリの `SKILL.md` frontmatter 由来**で、こちらの管理下にない。
    /// `../../../.claude` のような名前を `skillStore.appending(path:)` に渡すと
    /// FileManager がパスを解決するため、管理ルートの外へ書けてしまう。
    /// `assertMutable` は**削除・移動**しか守らないので、9 章のホワイトリストには
    /// 作成方向の穴が空く。ここがその穴を塞ぐ。
    ///
    /// 落とすのは危険なものだけで、文字種は絞らない — 実在するスキル名を
    /// 勝手に弾くと、正しい取得物が入らなくなる方の事故になる。
    public static func isValidName(_ name: String) -> Bool {
        guard !name.isEmpty, name.utf8.count <= 255 else { return false }
        guard name != ".", name != ".." else { return false }
        guard !name.hasPrefix(".") else { return false }        // 隠しファイルを作らせない
        guard !name.contains("/"), !name.contains("\\") else { return false }
        // `apps/web:deploy` の区切り。名前に含むと修飾名と区別できなくなる。
        guard !name.contains(":") else { return false }
        // 改行や NUL を含む名前は、表示にもパスにも使えない。
        guard !name.unicodeScalars.contains(where: {
            $0.properties.generalCategory == .control
        }) else { return false }
        return !isDenied(URL(filePath: name))
    }

    /// 通らなければ throw する。**作成の直前に必ず通す。**
    public static func assertValidName(_ name: String) throws {
        guard isValidName(name) else { throw Denial.invalidName(name) }
    }

    /// 削除・移動してよいか。通らなければ throw する。
    ///
    /// 許すのは 2 つだけ:
    ///   1. 自分が張った symlink（リンク先が実体置き場の配下）
    ///   2. registry.json に載っている実体（実体置き場 or 退避ディレクトリの配下）
    public static func assertMutable(
        _ url: URL, env: Environment, registry: Registry
    ) throws {
        try Registry.assertReadable(env: env)
        let path = url.standardized.path(percentEncoded: false)

        try assertNotBundled(url, env: env)
        if isDenied(url) { throw Denial.deniedPath(path) }

        if isSymlink(url) {
            guard let target = symlinkTarget(url),
                  env.managedRoots.contains(where: { isInside(target, $0) })
            else { throw Denial.symlinkOutsideStore(path) }
            return
        }

        guard let managedRoot = env.managedRoots.first(where: { isInside(url, $0) }) else {
            throw Denial.outsideManagedRoots(path)
        }
        try assertSafeCreation(url, inside: managedRoot, anchor: env.home)
        // Subagent は `<name>.md` なので拡張子を落とす。
        let name = url.pathExtension == "md"
            ? String(url.lastPathComponent.dropLast(3))
            : url.lastPathComponent
        let kind: Kind = url.pathExtension == "md" ? .subagent : .skill
        guard registry.entry(named: name, kind: kind) != nil else {
            throw Denial.notInRegistry(name)
        }
    }

    /// 作成・移動先の途中にある symlink を拒否する。文字列上の配下判定だけでは、
    /// `~/.agents/skills -> /outside` のような付け替えで管理外へ到達するため。
    public static func assertSafeCreation(_ url: URL, inside root: URL,
                                          anchor: URL? = nil) throws {
        let trusted = (anchor ?? root).standardizedFileURL
        let target = url.standardizedFileURL
        let boundary = root.standardizedFileURL
        guard target.path == boundary.path || isInside(target, boundary),
              boundary.path == trusted.path || isInside(boundary, trusted) else {
            throw Denial.outsideManagedRoots(target.path)
        }

        let trustedParts = trusted.pathComponents
        let targetParts = target.deletingLastPathComponent().pathComponents
        guard targetParts.starts(with: trustedParts) else {
            throw Denial.outsideManagedRoots(target.path)
        }
        var current = trusted
        for component in targetParts.dropFirst(trustedParts.count) {
            current.append(path: component)
            guard FileManager.default.fileExists(atPath: current.path) else { continue }
            if isSymlink(current) { throw Denial.unsafeParent(current.path) }
        }
    }

    /// Installed artifacts may be removed regardless of who installed them.
    /// Only a direct child of a known resource directory is eligible; never follow a link to delete its target.
    public static func assertUserArtifact(_ url: URL, kind: Kind, project: String? = nil,
                                          env: Environment) throws {
        guard kind == .skill || kind == .subagent else { throw Denial.deniedPath(url.path) }
        try assertNotBundled(url, env: env)
        if let project {
            let known = Set(ProjectScan.projectPaths(in: env).compactMap(ProjectScan.identity))
            guard let identity = ProjectScan.identity(project), known.contains(identity) else {
                throw Denial.outsideManagedRoots(url.path)
            }
        }
        let parent = url.deletingLastPathComponent().standardizedFileURL
        let base = (project.map { URL(filePath: $0) } ?? env.home).standardizedFileURL
        // `dropFirst` は parent が base の下にある前提でしか意味を持たない。
        // 外にあるパスを渡されると出鱈目な suffix が出るので、先に境界を確かめる。
        guard isInside(parent, base) || parent.path == base.path else {
            throw Denial.outsideManagedRoots(url.path)
        }
        let suffix = String(parent.path.dropFirst(base.path.count)).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let expected = base.resolvingSymlinksInPath().appending(path: suffix).standardizedFileURL
        guard isManagedParent(suffix, kind: kind, inProject: project != nil),
              parent.resolvingSymlinksInPath().path == expected.path,
              !url.lastPathComponent.hasPrefix("."), !isDenied(url),
              kind != .subagent || url.pathExtension == "md" else {
            throw Denial.outsideManagedRoots(url.path)
        }
    }

    /// 削除してよい親ディレクトリか（ホーム相対 / プロジェクト相対）。**純粋関数**（10.1）。
    ///
    /// **許可ルートを列挙し直さずに判定する。** 以前はプロジェクト側で
    /// `ProjectScan.skillRoots(project)` を呼んでいたが、あれはプロジェクト配下を
    /// 深さ 3 まで歩く I/O で、UI の `removableFiles` から **1 行ずつ**呼ばれていた
    /// （実測 0.16 秒 × 350 行 = 56 秒、しかも body 評価のたびにメインスレッド上で）。
    /// `ProjectScan.walk` が作りうる形をそのまま述語にすれば、同じ集合を歩かずに引ける。
    ///
    /// 判定する集合は変えない — プロジェクトのスキルはサブディレクトリの
    /// `.claude/skills` まで、サブエージェントは直下の `.claude/agents` だけ。
    static func isManagedParent(_ relative: String, kind: Kind, inProject: Bool) -> Bool {
        guard kind == .skill || kind == .subagent else { return false }
        guard inProject else {
            let labels = kind == .skill ? Agent.allCases.flatMap(\.skillRoots)
                                        : Agent.allCases.flatMap(\.subagentRoots)
            return labels.contains(relative) && !Agent.bundledSkillRoots.contains(relative)
        }
        let leaf = kind == .skill ? ".claude/skills" : ".claude/agents"
        if relative == leaf { return true }
        guard kind == .skill, relative.hasSuffix("/" + leaf) else { return false }
        return ProjectScan.isWalkablePrefix(String(relative.dropLast(leaf.count + 1)))
    }

    /// アプリ自身が作ったバックアップだけを消してよい（9 章）。
    ///
    /// `PermissionWriter` は世代を上限まで刈るために削除を要る唯一の場所で、
    /// `check-invariants.sh` の削除許可リストに載る 3 つ目のファイルになる。
    /// **許可リストに載せるだけでは中身が空でも通ってしまう**ので、
    /// `SkillManager` / `Updater` が `assertMutable` を必ず呼ぶのと同じ形で、
    /// ここを通すことを検査スクリプト側でも要求する。
    ///
    /// 許すのは自分の保存領域直下の**通常ファイル**だけ。ディレクトリと symlink は
    /// 辿った先を消しうるので拒否する。
    public static func assertAppBackup(_ url: URL, env: Environment) throws {
        let root = env.appSupport.appending(path: PermissionWriter.backupDirectory)
        guard isInside(url, root) else {
            throw Denial.outsideManagedRoots(url.standardized.path(percentEncoded: false))
        }
        guard !isSymlink(url) else {
            throw Denial.symlinkOutsideStore(url.standardized.path(percentEncoded: false))
        }
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(
                atPath: url.path(percentEncoded: false), isDirectory: &isDir),
              !isDir.boolValue else {
            throw Denial.deniedPath(url.standardized.path(percentEncoded: false))
        }
    }

    static func assertNotBundled(_ url: URL, env: Environment) throws {
        let protected = Agent.bundledSkillRoots.map { env.home.appending(path: $0) }
            + [env.home.appending(path: ".codex/plugins"), env.home.appending(path: ".claude/plugins"),
               env.home.appending(path: ".cursor/plugins")]
        let paths = [url.standardizedFileURL, url.resolvingSymlinksInPath()]
        guard !paths.contains(where: { path in protected.contains { path.path == $0.path || isInside(path, $0) } }) else {
            throw Denial.deniedPath(url.path)
        }
    }

    // MARK: - 判定

    static func isDenied(_ url: URL) -> Bool {
        deniedNames.contains(url.lastPathComponent)
            || deniedExtensions.contains(url.pathExtension)
    }

    static func isSymlink(_ url: URL) -> Bool {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path(percentEncoded: false))
        return attrs?[.type] as? FileAttributeType == .typeSymbolicLink
    }

    /// リンク先を絶対パスで返す。相対 symlink も解決する。
    static func symlinkTarget(_ url: URL) -> URL? {
        guard let dest = try? FileManager.default.destinationOfSymbolicLink(atPath: url.path(percentEncoded: false))
        else { return nil }
        let target = dest.hasPrefix("/")
            ? URL(filePath: dest)
            : url.deletingLastPathComponent().appending(path: dest)
        return target.standardized
    }

    /// `..` で抜けられないよう、パス境界を見て判定する。
    static func isInside(_ url: URL, _ root: URL) -> Bool {
        let path = url.standardized.path(percentEncoded: false)
        var rootPath = root.standardized.path(percentEncoded: false)
        if !rootPath.hasSuffix("/") { rootPath += "/" }
        return path.hasPrefix(rootPath)
    }
}

File: src/writeGuard.ts
import {
  cpSync, existsSync, linkSync, lstatSync, mkdirSync, mkdtempSync, readlinkSync, realpathSync,
  renameSync, rmSync, statSync, symlinkSync, unlinkSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, extname, isAbsolute, join, resolve, sep } from "node:path";
import { BUNDLED_SKILL_ROOTS, KindId } from "./agent";
import { SKILL_SOURCES, SUBAGENT_SOURCES } from "./source";
import { agentStore, disabledAgentStore, Env, managedRoots } from "./env";
import { AgentToolError } from "./errors";
import { assertReadable, entry, Registry } from "./registry";

/**
 * 触ってよい対象の限定。「触ってはいけないパスの列挙」ではなく
 * 「触ってよい対象の限定」にしてあるので、新しいリソース種別を足しても安全側に倒れる。
 */

/**
 * ホワイトリストを通っても無条件で拒否する。ホワイトリストの実装ミス 1 つで
 * 到達しうる場所なので二重にする。`~/.codex/auth.json` は誤爆すると全ログインが飛ぶ。
 */
export const DENIED_NAMES: ReadonlySet<string> = new Set([
  "auth.json", "oauth_creds.json", "settings.json",
  "settings.local.json", ".claude.json", "config.toml", "mcp.json",
]);
export const DENIED_EXTENSIONS: ReadonlySet<string> = new Set(["sqlite", "sqlite-wal", "sqlite-shm"]);

/** Windows のパスは大文字小文字を区別しないので、比較の前に畳む。 */
const fold = (value: string): string => process.platform === "win32" ? value.toLowerCase() : value;
const key = (path: string): string => fold(resolve(path));

/** `..` で抜けられないよう、パス境界を見て判定する。 */
export const isInside = (path: string, root: string): boolean =>
  key(path).startsWith(key(root).endsWith(sep) ? key(root) : key(root) + sep);

const isSame = (a: string, b: string): boolean => key(a) === key(b);

export const isDenied = (path: string): boolean =>
  DENIED_NAMES.has(basename(path)) || DENIED_EXTENSIONS.has(extname(path).slice(1));

/** symlink と Windows の junction はどちらもここで真になる。hardlink は「リンク」ではない。 */
export function isLink(path: string): boolean {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

/** リンク先を絶対パスで返す。相対リンクも解決する。 */
export function linkTarget(path: string): string | null {
  try {
    const destination = readlinkSync(path);
    return resolve(isAbsolute(destination) ? destination : join(dirname(path), destination));
  } catch {
    return null;
  }
}

const realOrSelf = (path: string): string => {
  try {
    return realpathSync(path);
  } catch {
    return resolve(path);
  }
};

/**
 * 取得物が名乗った名前を、そのままパス要素に使ってよいか。純粋関数。
 *
 * 名前は取得先リポジトリの frontmatter 由来でこちらの管理下にない。
 * `../../../.claude` のような名前を join に渡すと管理ルートの外へ書けてしまう。
 * `assertMutable` は削除・移動しか守らないので、作成方向の穴をここで塞ぐ。
 *
 * 落とすのは危険なものだけで、文字種は絞らない — 実在するスキル名を弾く方が実害になる。
 */
export function isValidName(name: string): boolean {
  if (name === "" || Buffer.byteLength(name, "utf8") > 255) return false;
  if (name === "." || name === "..") return false;
  if (name.startsWith(".")) return false;                    // 隠しファイルを作らせない
  if (name.includes("/") || name.includes("\\")) return false;
  if (name.includes(":")) return false;                      // `apps/web:deploy` の区切り
  if (/\p{Cc}/u.test(name)) return false;                    // 改行や NUL
  return !isDenied(name);
}

/** 通らなければ throw する。作成の直前に必ず通す。 */
export function assertValidName(name: string): void {
  if (!isValidName(name)) {
    throw new AgentToolError("INVALID_NAME", `${name} cannot be used as a name`);
  }
}

/**
 * 削除・移動してよいか。許すのは 2 つだけ:
 *   1. 自分が張ったリンク（リンク先が実体置き場の配下）
 *   2. registry に載っている実体（実体置き場 or 退避ディレクトリの配下）
 * Windows の Subagent は hardlink なのでリンクとして見えず、実体との同一性で判定する。
 */
export function assertMutable(path: string, env: Env, registry: Registry): void {
  assertReadable(env);
  const target = resolve(path);

  assertNotBundled(target, env);
  if (isDenied(target)) {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${target} is protected`);
  }

  if (isLink(target)) {
    const destination = linkTarget(target);
    if (!destination || !managedRoots(env).some(root => isInside(destination, root))) {
      throw new AgentToolError("SYMLINK_OUTSIDE_STORE", `${target} was not linked by Agent Tool`);
    }
    return;
  }

  // Subagent は `<name>.md` なので拡張子を落とす。
  const isSubagent = extname(target) === ".md";
  const name = isSubagent ? basename(target, ".md") : basename(target);
  const kind: KindId = isSubagent ? "subagent" : "skill";

  const root = managedRoots(env).find(candidate => isInside(target, candidate));
  if (root) {
    // 退避ディレクトリは appSupport の下にあり、ホームの外に置かれることがある。
    assertSafeCreation(target, root, isInside(root, env.home) ? env.home : env.appSupport);
  } else if (!isHardLinkedIntoStore(target, name, env)) {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${target} is outside the managed roots`);
  }

  if (!entry(registry, name, kind)) {
    throw new AgentToolError("NOT_IN_REGISTRY", `${name} is managed by another tool`);
  }
}

/** Windows の Subagent は hardlink で配る。実体と同じ inode なら自分が張ったもの。 */
function isHardLinkedIntoStore(path: string, name: string, env: Env): boolean {
  if (process.platform !== "win32" || extname(path) !== ".md") return false;
  try {
    const link = statSync(path);
    return [agentStore(env), disabledAgentStore(env)].some(store => {
      try {
        const source = statSync(join(store, `${name}.md`));
        return source.ino === link.ino && source.dev === link.dev;
      } catch {
        return false;
      }
    });
  } catch {
    return false;
  }
}

/**
 * 作成・移動先の途中にあるリンクを検査する。文字列上の配下判定だけでは、
 * `~/.agents/skills -> /outside` のような付け替えで管理外へ到達するため。
 *
 * リンクそのものは拒まない。`~/.claude` や `~/.agents` を dotfiles リポジトリへ張るのは
 * 普通の構成で、一律に弾くと有効化も更新もできなくなる。
 * 拒むのは信頼できる根（ホーム / プロジェクト）の外へ出るリンクだけ。
 */
export function assertSafeCreation(path: string, root: string, anchor?: string): void {
  const trusted = resolve(anchor ?? root);
  const boundary = resolve(root);
  const target = resolve(path);
  const deny = (): never => {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${target} is outside the managed roots`);
  };
  if (!(isSame(target, boundary) || isInside(target, boundary))) deny();
  if (!(isSame(boundary, trusted) || isInside(boundary, trusted))) deny();

  const trustedParts = trusted.split(sep);
  const targetParts = dirname(target).split(sep);
  if (!trustedParts.every((part, i) => fold(part) === fold(targetParts[i] ?? "\u0000"))) deny();

  // リンク先は realpath で比較する。macOS の `/var` のように、根そのものが
  // symlink を含むことがあり、片側だけ解決すると同じ場所を別物と判定する。
  const trustedReal = realOrSelf(trusted);
  let current = trusted;
  for (const component of targetParts.slice(trustedParts.length)) {
    current = join(current, component);
    if (!isLink(current)) continue;
    const destination = linkTarget(current);
    const resolved = destination === null ? null : realOrSelf(destination);
    if (!resolved || !isInside(resolved, trustedReal)) {
      throw new AgentToolError("WRITE_GUARD_DENIED",
        `${current} is a symbolic link that leaves the trusted root`);
    }
    current = resolved;
  }
}

/**
 * 他のツールが入れた実体の置き場。走査ホワイトリストのうち、
 * ホーム直下の既知ルートから同梱ルートを除いたもの。
 */
export function userRoots(kind: KindId, env: Env): string[] {
  const sources = kind === "skill" ? SKILL_SOURCES : SUBAGENT_SOURCES;
  return sources.flatMap(source => {
    if (source.kind === "cli" || source.root !== "home") return [];
    return BUNDLED_SKILL_ROOTS.has(source.path) ? [] : [join(env.home, source.path)];
  });
}

/**
 * 他のツールが入れた実体を触ってよいか。registry には載っていないが、
 * 走査ホワイトリストの既知ルート**直下**にあるものだけを許す。
 *
 * 「誰が入れたか」ではなく「どこに在るか」で決める — 利用者から見れば
 * 一覧に出ているものは自分の資産で、入れた経路が違うだけで消せないのは通らない。
 * ただしリンクは辿らない。辿ると、リンク先の管理外ファイルまで消しにいく。
 */
export function assertUserArtifact(path: string, kind: KindId, env: Env): void {
  assertArtifact(path, kind, userRoots(kind, env), env.home, env);
}

/** プロジェクト内の Skill / Subagent の置き場。一覧が読む場所と同じにする。 */
export const projectRoots = (kind: KindId, project: string): string[] =>
  [join(project, ".claude", kind === "subagent" ? "agents" : "skills")];

/**
 * プロジェクト内の実体を触ってよいか。信頼の根はワークスペースで、
 * `.claude/skills` / `.claude/agents` の**直下**だけを許す。
 * ホームの管理ルートとは別の根なので、`assertMutable` ではなくこちらを通す。
 */
export function assertProjectArtifact(path: string, kind: KindId, project: string, env: Env): void {
  assertArtifact(path, kind, projectRoots(kind, project), project, env);
}

function assertArtifact(path: string, kind: KindId, roots: string[], anchor: string, env: Env): void {
  if (kind !== "skill" && kind !== "subagent") {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${kind} is managed by the agent`);
  }
  const target = resolve(path);
  assertNotBundled(target, env);
  if (isDenied(target) || basename(target).startsWith(".")) {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${target} is protected`);
  }
  if (kind === "subagent" && extname(target) !== ".md") {
    throw new AgentToolError("WRITE_GUARD_DENIED", `${target} is not a subagent file`);
  }
  const parent = dirname(target);
  if (!roots.some(root => isSame(parent, root))) {
    throw new AgentToolError("WRITE_GUARD_DENIED",
      `${target} is not directly inside a known ${kind} root`);
  }
  // 既知ルートまでの経路が信頼の根の外へ張り替えられていないことを確かめる。
  assertSafeCreation(target, parent, anchor);
}

/** 同梱スキルとプラグインは、誰が入れたものでも触らない。 */
export function assertNotBundled(path: string, env: Env): void {
  const protectedRoots = [
    ...[...BUNDLED_SKILL_ROOTS].map(root => join(env.home, root)),
    join(env.home, ".codex", "plugins"),
    join(env.home, ".claude", "plugins"),
    join(env.home, ".cursor", "plugins"),
  ];
  const candidates = [resolve(path), realOrSelf(path)];
  const hit = candidates.some(candidate =>
    protectedRoots.some(root => isSame(candidate, root) || isInside(candidate, root)));
  if (hit) throw new AgentToolError("WRITE_GUARD_DENIED", `${path} is protected`);
}

// MARK: - 書き込み
// ここから下だけが実体とリンクを触る。呼び出し側は先に `assertMutable` /
// `assertValidName` を通し、作成系は `prepare` で作成先を検査してから使う（設計決定 D-9）。

/** 作成先を検査してから親ディレクトリを用意する。 */
export function prepare(path: string, root: string, anchor?: string): void {
  assertSafeCreation(path, root, anchor);
  mkdirSync(dirname(path), { recursive: true });
}

/**
 * 実体へのリンクを張る。Windows では symlink に昇格が要るので、
 * ディレクトリは junction、ファイルは hardlink を使う（設計決定 D-4）。
 */
export function createLink(target: string, path: string, kind: "dir" | "file"): void {
  if (process.platform !== "win32") {
    symlinkSync(target, path);
    return;
  }
  if (kind === "dir") {
    symlinkSync(target, path, "junction");
    return;
  }
  try {
    linkSync(target, path);
  } catch (error) {
    // hardlink は同一ボリューム内でしか張れない。コピーへは落とさない —
    // 実体が 2 つになると、更新も無効化も片方にしか効かなくなる。
    throw new AgentToolError("OPERATION_FAILED",
      `${path} could not be hard-linked to ${target} (different volume?): ${(error as Error).message}`);
  }
}

/**
 * その位置が実体への「自分が張ったリンク」か。
 * Windows の hardlink はリンクとして見えないので、実体との同一性で判定する。
 */
export function isManagedLink(path: string, target: string): boolean {
  if (isLink(path)) {
    const destination = linkTarget(path);
    return destination !== null && isSame(destination, target);
  }
  if (process.platform !== "win32") return false;
  try {
    const a = statSync(path), b = statSync(target);
    return a.ino === b.ino && a.dev === b.dev && a.ino !== 0;
  } catch {
    return false;
  }
}

/** リンクだけを外す。実体は消さない。 */
export function removeLink(path: string): void {
  unlinkSync(path);
}

/** 実体を消す。D-5 によりゴミ箱へは送らないので、呼び出し前に必ず確認を取る。 */
export function remove(path: string): void {
  rmSync(path, { recursive: true, force: true });
}

export function move(from: string, to: string): void {
  try {
    renameSync(from, to);
  } catch (error) {
    // ボリュームをまたぐと rename は EXDEV で失敗する（退避先が別ドライブのとき）。
    if ((error as NodeJS.ErrnoException).code !== "EXDEV") throw error;
    cpSync(from, to, { recursive: true, verbatimSymlinks: true });
    rmSync(from, { recursive: true, force: true });
  }
}

export function copy(from: string, to: string): void {
  cpSync(from, to, { recursive: true, verbatimSymlinks: true });
}

export const exists = (path: string): boolean => existsSync(path);

/** OS の一時領域に作業用ディレクトリを作る。呼び出し側が全経路で `remove` する。 */
export const stagingDir = (prefix: string): string => mkdtempSync(join(tmpdir(), prefix));
